### DISCUSSION TITLE
Send hex to rs485

### DISCUSSION THREAD

**Discussion by axel-GB** [arendst/Tasmota#22653] [Q&A] [answered]
I have configured TCP TX, RX and runing TCPStart 8888. 
There is Rs485 connected. 
Conection from other devices works. 

I have a problem with sending a command in Hex from the same TASMOTA runing TCP server to RS.
 Command like SerialSend5 FFFFC10FDE
Tested  Websend [192.168.1.2:8888] FFFFC10FDE but not working

**Comment by sfromis** [comment]
When you hand over serial to another device via TCP (ser2net protocol), you can no longer use local serial handling for the same. Your other device has control and that's where you can send and so on. Trying to use http (via WebSend) is badly useless for ser2net, and if it even made sense, it would still not be viable on the same device.

**Reply by axel-GB** [reply]
OK how can i send command from other Tasmota device.
Now I cant send anything from other device in HEX. 

ps.. I need to send TIME  every 20 minutes to rs485 in HEX, but

**Reply by sfromis** [reply]
Again, when you hand over control over a serial port to another device, that's where everything has to be done. Configuring for serial via TCP means that you give up local access to serial.

**Reply by axel-GB** [reply]
YES. I will send command from other device - Tasmota.
But how, when i send 
Websend [192.168.1.2:8888] FFFFC10FDE
on server i get 

to MCU/2: 474554202F636D3F636D6E643D4646464643313046444520485454502F312E310D0A486F73743A203139322E3136382E352E3234313A383838380D0A557365722D4167656E743A20455350333248545450436C69656E740D0A436F6E6E656374696F6E3A206B6565702D616C6976650D0A4163636570742D456E636F64696E673A206964656E746974793B713D312C6368756E6B65643B713D302E312C2A3B713D300D0A0D0A

Question is how send Hex command to Server? To get FFFFC10FDE

**Reply by sfromis** [reply]
I already said that it is badly useless to even try to mix WebSend and serial. The hex data you posted now decodes to:
```
GET /cm?cmnd=FFFFC10FDE HTTP/1.1
Host: 192.168.5.241:8888
User-Agent: ESP32HTTPClient
Connection: keep-alive
Accept-Encoding: identity;q=1,chunked;q=0.1,*;q=0
```
Probably not what you wanted.

More generally, you are asking without describing any details of what exactly you mean by words like "server", "send" and "command", and how the communication is configured and setup among what types of devices. I'm not trying to be difficult, but the things you post make so little sense that it is hard to guess what you talk about.

**Reply by axel-GB** [reply]
I have a Acces Controler connected by RS485 to Serial with Tasmota as TCP Serwer.
I need send command to set TIME for working with TIMERS. 
But Now for test I try to open door (it works with android UCP/TCP Widget APP).

Really i send websend [192.168.5.241:8888] FFFFC10FDE

Command you decoded GET /cm?cmnd=FFFFC10FDE HTTP/1.1    is good 
Host: 192.168.5.241:8888  you decoded is  CORRECT    ( It has been hanged by my phone and changed it - soory)
but that what i send isn't what i get and need to correct working.
How have you decoded it ????


Data decoded was correct but do not works to open door. If i get FFFFC10FDE in console it should work, like the android app do.

**Reply by sfromis** [reply]
Again, it does not make sense to try to use WebSend for serial. Zero sense. Waste of time.

Forgetting everything about WebSend - what command would you use on the Tasmota "server" to send the serial data you want? Maybe something like `SerialSend5 xxxxxx` ? 

That would be the command you *could* send with `WebSend [92.168.5.241] command` 
Of course make sure to use the IP (or host name) of the Tasmota device owning the serial port *without* having a config for TCP serial tunnel, and absolutely not a port number used for such tunneling. In general, using MQTT is the preferred strategy for communicating among Tasmota devices, but `WebSend` can usually also work.

My decoding was just a simple tool to "convert hex to ascii", you can find other such tools by Googling for that phrase.

**Reply by axel-GB** [reply]
SerialSend does not work to send anything and anywhere.

But from oder device sended
websend [192.168.5.241] FFFFC10FDE
i get:

07:50:51.140 SRC: WebCommand from 192.168.5.220
07:50:51.143 CMD: Grp 0, Cmd 'FFFFC10FDE', Idx 1, Len 0, Pld -99, Data ''
07:50:51.148 RSL: RESULT = {"Command":"Unknown","Input":"FFFFC10FDE"}
and do not work
Please help.
PS. it can be from berry script to because i will nedd to make hex conversion.

**Reply by sfromis** [reply]
As Tasmota has no command named "FFFFC10FDE", it does not help submitting such a bad command.

That's why I suggested skipping all that talk about WebSend, and make it clear what Tasmota console command you'd use to send those hex bytes, on a Tasmota device owning the serial port **without** it being configured for TCP.

**If** that is something with `SerialSend5` (that was not a suggestion, merely an example), **that** would be the command you can use with `WebSend`. But you need a valid Tasmota command, until you have that, using WebSend does not make sense.

If you do it on a Tasmota device with a serial port configured for TCP, you cannot do **anything** with that serial port, except from the device you have given control over it to.

**Reply by barbudor** [reply]
I believe @axel-GB is confused and mixing up 2 different modes of Tasmota
- TcpBridge which is enabled by setting the GPIOs to TcpRX and TcpRX. In that mode Tamsota bridges a TCP port (exemple 8888 and the serial). To use that mode, once you have performed `TcpStart 8888` you have to open a socket to port 8888 from another software, most probably a computer and use the socket interface to send binary data (not hex). That is not using the HTTP protocol anywhere. You MUST write a software that open the socket. Any programming language would work (python, C/C++ or even Berry on another Tasmota)

- SerialBridge is enabled using SerBrTx and SerbrRX where Tasmota can send and receive data to/from the serial ports using Tasmota commands such as `SerialSend5 FFFFC10FDE`. Such commands can be executed from the web console, from rules or sent to Tasmota using MQTT or HTTP (`https://ip.of.your.tasmota?cmnd=SerialSend5%20FFFFC10FDE`) or from another Tasmota device using the WebSend command such as `WebSend [ip.of.your.tasmita] SerialSend5 FFFFC10FDE`

So try to understand the differences between both and make your mind up on which mode you want to use

**Comment by axel-GB** [comment]
Thanks @barbudor.
You write 
"  You MUST write a software that open the socket. Any programming language would work (python, C/C++ or even Berry on another Tasmota)"


 If I have understood that is what I need to do  - Open socket in  Berry and send data from other Tasmota.

**Reply by sfromis** [reply]
Such a Berry program should adhere to the ser2net protocol, which I assume is more than just sending the data you want to get across.

**Reply by barbudor** [reply]
@axel-GB if you just need to send that simple hex string, why not using SerialBridge
It will be far more simple than TcpBridge

**Reply by axel-GB** [reply]
Thanks to all.
Sending works in Berry.

I need help with string.format %x it want work.
When I get 
month = tasmota.time_dump(tasmota.rtc()['local'])['month']
month is 12
print(string.hex(month))
I get "C" and i need send two digits like "0C"

**Reply by joba-1** [reply]
this probably works like in C’s printf. Try %02x

**Reply by s-hadinger** [reply]
Indeed. Use string.format with %02X

**Reply by sfromis** [reply]
Simplest is to use `f'{month:02X}'` giving `0C`. Or lower-case x for lower-case hex.

**Reply by axel-GB** [reply]
Thanks @sfromis , @s-hadinger .
It works. And what is the simples way to get from 2024 only 2 digits 24
year = tasmota.time_dump(tasmota.rtc()['local'])['year']

Because I use 
def strReplace(haystack, needle, food)
  import string
  return string.split(haystack, needle).concat(food)
end

**Reply by sfromis** [reply]
Right now, `year % 100 == 24`

Not sure what the related use case of the last part with `strReplace` is, but usually it is preferable to construct the string as wanted, and possibly placeholders. Like:
```
var year2 = tasmota.time_dump(tasmota.rtc('local'))['year'] % 100
print(f'Current year is {year2:02i} in two dec digits')
print(f'Current year is {year2:02X} in two hex digits')
```
(Also note that using `tasmota.rtc` can be a bit shorter, by only requesting the `local` value).

**Reply by axel-GB** [reply]
Thanks. 
Marry X-mas.

**Reply by axel-GB** [reply]
Could You tell me where the problem is
`
def tim()
    if year == 1970
    tc = 0
    print("sprawdzam rok")
    czas()
    print("rok", year)
    tasmota.set.timer(500,tim)
    else 
       if year > 1970 && tc ==0
        hexsend2.data()
       tc =1
       end
    end
end
#...
class MyClass
     def every_second() 
tim()
end
end
d4 = MyClass()
tasmota.add_driver(d4)
`
I get:

1970
sprawdzam rok
rok 2024
BRY: Exception> 'attribute_error' - the 'Tasmota' object has no attribute 'set'

all works but the attribute error

**Reply by sfromis** [reply, marked as answer]
Typo, `tasmota.set.timer` should have been `tasmota.set_timer`

**Reply by axel-GB** [reply]
Thanks @sfromis

**Reply by axel-GB** [reply]
Hi.
Is there any way to get (in Berry) ESP device ID ?
Mac address last 4 bytes.

**Reply by sfromis** [reply]
`tasmota.wifi('mac')` gives you a string with the MAC. Note that the MAC is different in AP vs STA mode, and if you use Ethernet instead of Wifi. IIRC, the Wifi MAC is the one closest to the device id.

FTR, the ESP device id is the last 3 bytes of the binary MAC, as the first 3 is to identify the organization allocating numbers within the 24-bit range of the last 3.

**Reply by axel-GB** [reply]
Thx. 
Will the Mac address change after (re)flashing the device?

**Reply by sfromis** [reply]
In that case it would not be relevant as base for a "device id".

**Comment by axel-GB** [comment]
Why it want work?
name = input("Please enter your name: ")
print("Hello, " + name + "! Good to see you.")


age = int(input("Please enter your age: "))
print("Wow, you are " + str(age) + " years old!")

Please enter your age: Wow, you are 0 years old!

**it doesn't wait for input**. 

Second question  how can I make visible menu with select option 1 Name 2 data to get input

**Reply by sfromis** [reply]
The `input` function of the Berry language is not supported in the Tasmota integration, as there is no terminal connection. And the Berry REPL console is only meant for light testing of small bits of code, not as a place to actually use programs.

Here's an example of a Berry program setting up a web UI page with input fields.
https://github.com/arendst/Tasmota/blob/development/tasmota/berry/artnet/artnet_ui.be
